### PR TITLE
`podman cp` supports .dockerignore

### PR DESCRIPTION
**Pull request by QiWang19**
podman cp can exclude (or keep) some files by specifying the rules in .dockerignoe file, so that can avoid sending unnecessary files to container.

Signed-off-by: Qi Wang <qiwan@redhat.com>

### CODE REVIEW

**Comment by mheon** [comment]
Alright, this is getting ridiculous, we've had to manually whitelist every PR from Qi...

**Comment by mheon** [comment]
bot, add author to whitelist

**Comment by TomSweeneyRedHat** [comment]
Ditto on Buildah, if you find the magic bits to flip @mheon, please let me know.

**Comment by QiWang19** [comment]
Because my format vendor PR ruined the libpod repo last week :fearful: ? bot, I'm sorry :sob:

**Comment by rhatdan** [comment]
I think we have to add her to the OWNERS file.

File: cmd/podman/cp.go [hunk 1]
```diff
@@ -68,11 +76,13 @@ func cpCmd(c *cliconfig.CpValues) error {
 	}
 	defer runtime.Shutdown(false)
 
-	extract := c.Flag("extract").Changed
-	return copyBetweenHostAndContainer(runtime, args[0], args[1], extract)
+	copyOpts := copyOptions{
+		extract: c.Flag("extract").Changed,
```
Comment (rhatdan) [inline cmd/podman/cp.go]: This code is wrong
Should be 
```
copyOpts := copyOptions{
		extract:  c.Extract 
}
```

File: cmd/podman/cp.go [hunk 2]
```diff
@@ -253,11 +272,58 @@ func copy(src, destPath, dest string, idMappingOpts storage.IDMappingOptions, ch
...
+				if err != nil {
+					return err
+				}
+				if !match {
+					continue
+				}
+				if exclude.IsExcluded {
+					return nil
+				}
+				break
```
Comment (giuseppe) [inline cmd/podman/cp.go]: why `break` here?  Don't we need to check all the exclude rules?

File: cmd/podman/cp.go [hunk 3]
```diff
@@ -253,11 +272,58 @@ func copy(src, destPath, dest string, idMappingOpts storage.IDMappingOptions, ch
...
+		if err != nil {
+			return err
+		}
+		if !match {
+			continue
+		}
+		if exclude.IsExcluded {
+			return nil
+		}
+		break
```
Comment (giuseppe) [inline cmd/podman/cp.go]: same here, don't we need to check all the excludes?

File: cmd/podman/cp.go [hunk 4]
```diff
@@ -253,11 +272,58 @@ func copy(src, destPath, dest string, idMappingOpts storage.IDMappingOptions, ch
...
+				return errors.Wrapf(err, "error copying %q to %q", srcPath, dest)
+			}
+			return nil
+		}
+		err := filepath.Walk(srcAbsPath, func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			if info.IsDir() {
+				return nil
```
Comment (giuseppe) [inline cmd/podman/cp.go]: what happens if a directory is marked to be excluded?

For example, if I've:
```
mkdir -p a/b
touch a/b/c

echo a/b > .dockerignore
```

Shouldn't we skip also the file `c` from being copied?

Comment (QiWang19) [reply cmd/podman/cp.go]: fixed

File: cmd/podman/cp.go (hunk 2 above)
Comment (QiWang19) [reply cmd/podman/cp.go]: if it's not match, I have `continue` in line 295

File: cmd/podman/cp.go (hunk 1 above)
Comment (QiWang19) [reply cmd/podman/cp.go]: fixed

File: cmd/podman/cp.go (hunk 2 above)
Comment (giuseppe) [reply cmd/podman/cp.go]: shouldn't we consider the last match though?

Something like:
```
a/b
!a/b/c
```
Wouldn't we skip `a/b/c` in this way?

Comment (QiWang19) [reply cmd/podman/cp.go]: yes, @giuseppe , needs review :male_detective:

File: cmd/podman/cp.go [hunk 5]
```diff
@@ -68,11 +76,13 @@ func cpCmd(c *cliconfig.CpValues) error {
 	}
 	defer runtime.Shutdown(false)
 
-	extract := c.Flag("extract").Changed
-	return copyBetweenHostAndContainer(runtime, args[0], args[1], extract)
+	copyOpts := copyOptions{
+		extract: c.Extract,
```
Comment (mheon) [inline cmd/podman/cp.go]: Why are we only filling one field in here?

If we're not using all of `copyOptions`, just pass in Extract to `copyBetweenHostAndContainer` and make a `copyOptions` in there

File: cmd/podman/cp.go [hunk 6]
```diff
@@ -164,17 +174,23 @@ func copyBetweenHostAndContainer(runtime *libpod.Runtime, src string, dest strin
 	if len(glob) == 0 {
 		glob = append(glob, srcPath)
 	}
+
+	dir, err := os.Getwd()
```
[the remark this answers is not in the document]
Comment (QiWang19) [reply cmd/podman/cp.go]: [docker looks for a file named .dockerignore in the root directory of the context](https://docs.docker.com/engine/reference/builder/#dockerignore-file)
what does this mean, it it the source directory?

[the remark this answers is not in the document]
Comment (mheon) [reply cmd/podman/cp.go]: That's for Dockerfiles. We don't have a context directory in `podman copy`, really. We have a source and a destination we're copying.

[the remark this answers is not in the document]
Comment (mheon) [reply cmd/podman/cp.go]: I'm looking for documentation here on how `docker cp` actually handles this and not finding much.

[the remark this answers is not in the document]
Comment (mheon) [reply cmd/podman/cp.go]: Does Docker even do this? I can't find any documentation for this actually working in Docker. I don't really see why we would want to implement it if they don't?

[the remark this answers is not in the document]
Comment (QiWang19) [reply cmd/podman/cp.go]: I didn't find the documentation, either. But just found the cp command was influenced if there is a .dockerignore file in the context directory

[the remark this answers is not in the document]
Comment (QiWang19) [reply cmd/podman/cp.go]: hm...maybe we don't need this.

**Pull request closed without merging**